proof/refine/ARM/Syscall_R.thy

```diff
@@ -740,98 +740,69 @@ lemma isReply_awaiting_reply':
   "isReply st = awaiting_reply' st"
   by (case_tac st, (clarsimp simp add: isReply_def isBlockedOnReply_def)+)
 
-lemma doReply_invs[wp]:
-  "\<lbrace>tcb_at' t and reply_at' reply and invs' and sch_act_simple\<rbrace>
-   doReplyTransfer t reply grant
+lemma handleTimeout_invs':
+  "\<lbrace>invs' and st_tcb_at' active' tptr and sch_act_not tptr and ex_nonz_cap_to' tptr\<rbrace>
+   handleTimeout tptr timeout
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  apply (clarsimp simp: handleTimeout_def)
+  apply wpsimp
+      apply (rename_tac tcb)
+      apply (rule_tac Q="\<lambda>_. invs'"
+                  and E="\<lambda>_. invs' and st_tcb_at' active' tptr and sch_act_not tptr
+                             and (\<lambda>s. False \<longrightarrow> bound_sc_tcb_at' (\<lambda>a. a \<noteq> None) tptr s)
+                             and ex_nonz_cap_to' tptr
+                             and (\<lambda>s. \<exists>n\<in>dom tcb_cte_cases. cte_wp_at' (\<lambda>cte. cteCap cte
+                                                                               = cteCap (tcbTimeoutHandler tcb))
+                                                                        (tptr + n) s)"
+                   in hoare_post_impErr)
+        apply (rule sfi_invs_plus')
+       apply (wpsimp wp: getTCB_wp
+                   simp: isValidTimeoutHandler_def)+
+  apply (clarsimp simp: cte_wp_at'_obj_at' tcb_cte_cases_def  projectKOs obj_at'_def)
+  done
+
+crunches isValidTimeoutHandler
+  for inv: P
```

Comment:
nitpick: will it be ok, and better, to put this into `wp`?

Reply:
Maybe I'm being paranoid, and it's fine to put it into the `wp` set, but my thinking was that this rule throws away information about whether it *was* a valid timeout handler, and I didn't want to do that

Reply:
You're right to be paranoid, but the rule that ends up being generated has `λ_. P` as the postcondition, so it can only be applied when the return value is not being considered. In my view, this means that it actually is safe to be in the `wp` set. 

There is then a theoretical alternative rule that has `⦃P⦄` as the postcondition that would need to include that information.